python/ray/dag/compiled_dag_node.py

```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
 
         TODO(rui): catch the case that user holds onto the CompiledDAGRefs
         """
-        from ray.dag import DAGContext
+        if self._max_execution_index < execution_index:
+            from ray.dag import DAGContext
 
-        ctx = DAGContext.get_current()
-        if timeout is None:
-            timeout = ctx.retrieval_timeout
+            ctx = DAGContext.get_current()
+            if timeout is None:
+                timeout = ctx.retrieval_timeout
 
-        while self._max_execution_index < execution_index:
-            if self._max_execution_index + 1 == execution_index:
-                # Directly fetch and return without buffering
+            while self._max_execution_index < execution_index:
+                if len(self._result_buffer) >= self._max_buffered_results:
+                    raise ValueError(
+                        "Too many buffered results: the allowed max count for "
+                        f"buffered results is {self._max_buffered_results}; call "
+                        "ray.get() on previous CompiledDAGRefs to free them up "
+                        "from buffer."
+                    )
                 self._max_execution_index += 1
-                return self._dag_output_fetcher.read(timeout)
-            # Otherwise, buffer the result
-            if len(self._result_buffer) >= self._max_buffered_results:
-                raise ValueError(
-                    "Too many buffered results: the allowed max count for "
-                    f"buffered results is {self._max_buffered_results}; call ray.get() "
-                    "on previous CompiledDAGRefs to free them up from buffer."
-                )
-            self._max_execution_index += 1
-            start_time = time.monotonic()
-            self._result_buffer[
-                self._max_execution_index
-            ] = self._dag_output_fetcher.read(timeout)
-            if timeout != -1:
-                timeout -= time.monotonic() - start_time
-                timeout = max(timeout, 0)
+                start_time = time.monotonic()
+
+                # Fetch results from each output channel up to execution_index and store
+                # them separately to enable individual retrieval
+                for chan, fetcher in self._dag_output_fetchers.items():
+                    if self._max_execution_index not in self._result_buffer:
```

Comment:
can you just use

```
self._result_buffer = defaultdict(dict)
```

and we can remove this code

Reply:
Sure thing, will do!